Variant generic unified main

artlog opened the pull request:
Variant  Generic adapted/merged to main
Following  #4199

Variant generic is another choice than Poll type where chosen_rank is a json array of possible configurable text answers from poll creator/owner.

creating 

<img width="388" height="543" alt="Capture d’écran du 2025-08-02 23-47-44" src="https://github.com/user-attachments/assets/66512d30-bbdc-4d1f-99bf-fdb4bd2f3d48" />

selecting possible answers at creation time in configuration tab :

<img width="540" height="325" alt="image" src="https://github.com/user-attachments/assets/75c8f0ef-df6c-47ad-83b3-28edea1b3c8f" />
 
then polls are selectable

Currently used to be exported and computed outside nextcloud poll with csv, results are not visible nor computed in poll.

VoteVariant type in front was renamed VotingVariant to match php code naming. Choice to use VoteVariant into php could have been done too, prefering to keep api side naming. Unification seems needed for Api poll calls.

Hoping this is clearer now.

Commit:
feat: add rank voting feature merge chosenrank with 8.1.4 refactorings

reworked from
git@github.com:vinimoz/polls.git rank-vote-feature-vue3
TRAPPE Vincent <trappe.vincent@laposte.Net>

voting variant generic

a configuration list of answers for a vote.

merged lib/Db/Pool.php with main manually

Signed-off-by: philippe lhardy <philippe.lhardy@astrolabe.coop>

Commit:
hide count/result for non simple votingVariant

Signed-off-by: philippe lhardy <philippe.lhardy@astrolabe.coop>

Commit:
merge generic variant in main ( after 8.2.0rc.2)

Move code VoteItem code into VoteButton
Add drop-down list in VoteButton
adapt with new poll types

Signed-off-by: philippe lhardy <philippe.lhardy@astrolabe.coop>

Commit:
lint github warnings

Signed-off-by: philippe lhardy <philippe.lhardy@astrolabe.coop>

Commit:
reset VoteIndicator and use VoteItem

remove unused code in VoteIndicator
reset VoteIndicator handles only simple variant
move generic variant display in VoteItem

Signed-off-by: philippe lhardy <philippe.lhardy@astrolabe.coop>

Commit:
apply github review

address remarks from #4199

Signed-off-by: philippe lhardy <philippe.lhardy@astrolabe.coop>

Commit:
js/ts/vue/php lint - no functional change -

npm run format:fix
make cs-fix

Signed-off-by: philippe lhardy <philippe.lhardy@astrolabe.coop>

artlog commented:
To have a coherent functionality quickly i study this addition

Voting variant generic is disabled by default and should be activated globally by administrator. So there is no visible change at upgrade without an explicit action to enable variant.

Commit:
merge main, review fix

add a lost comment during merge

Signed-off-by: philippe lhardy <philippe.lhardy@astrolabe.coop>

Commit:
Poll factory for test with chosenRank

Signed-off-by: philippe lhardy <philippe.lhardy@astrolabe.coop>

artlog commented:
i just aligned existing code with current main and test request, no new change.

Current code add variant generic for which usage is for collecting votes only.
Results should be exported and computed outside nextcloud poll with csv, results are not visible nor computed in poll.

Majority judgement new method won't be done within this merge, but this will be required to provide it,

This does not include creation of new variant disabled by default from my prior comment, i will notify when done.

dartcafe commented:
@artlog If you make a rebase: TableSchema has moved to a version sub dir. I fear you have to solve this conflict manually. As far as I can  see, the only change is the addition of your new field to polls.

Commit:
variant creation selection is disabled by default

To be able to select variant in poll creation option should be checked
within user settings
allow to have this experimental functionality without impacting classical use

Signed-off-by: philippe lhardy <philippe.lhardy@astrolabe.coop>

dartcafe commented on lib/Migration/V2/TableSchema.php:
```diff
@@ -247,6 +247,7 @@ abstract class TableSchema {
 			'access' => ['type' => Types::STRING, 'options' => ['notnull' => true, 'default' => 'private', 'length' => 1024]],
 			'anonymous' => ['type' => Types::BIGINT, 'options' => ['notnull' => true, 'default' => 0, 'length' => 20]],
 			'allow_maybe' => ['type' => Types::BIGINT, 'options' => ['notnull' => true, 'default' => 1, 'length' => 20]],
+			'chosen_rank' => ['type' => Types::TEXT, 'options' => ['notnull' => true, 'default' => 1, 'length' => 200]],
```
Default should be '' not 1

artlog replied:
done

dartcafe commented on src/components/Configuration/ConfigRankOptions.vue:
```diff
@@ -0,0 +1,113 @@
+<template>
```
Add license header:
```
<!--
  - SPDX-FileCopyrightText: 2025 Nextcloud contributors
  - SPDX-License-Identifier: AGPL-3.0-or-later
-->

```

Commit:
.vue lint fix

npm run format:fix

Signed-off-by: philippe lhardy <philippe.lhardy@astrolabe.coop>

Commit:
pull request checks fix

- add default string value for TEXT
- add missing license header

Signed-off-by: philippe lhardy <philippe.lhardy@astrolabe.coop>

dartcafe commented:
Let's see, what CI says. :wink:

dartcafe commented:
> it is a little disapointing... what that's default problem.

I am wondering, too. I'll have a look at it. I don't know, what I don't see.

dartcafe commented:
>  looks like all other Types:TEXT are nullable... ( 'notnull' => false, 'default' => null ) while this one is not. will see tomorrow.

Umpf. :see_no_evil:  Yes, I think this is the problem, because a text field with default `''` is treated as `null`. If I remember right, it is because of Oracle DB, which has this limitation.

Commit:
chosen_rank Types:TEXT empty string nullable

Signed-off-by: philippe lhardy <philippe.lhardy@astrolabe.coop>

Commit:
remove chosenRank type, keep only column

Signed-off-by: philippe lhardy <philippe.lhardy@astrolabe.coop>

Commit:
support nullable chosenRank

- don't parse chosenRank for non generic variant
- delegate to existing getChosenRank of getters
- set default table column chosen_rank to null
- set default chosenRank to json empty array

Signed-off-by: philippe lhardy <philippe.lhardy@astrolabe.coop>

artlog commented on src/components/VoteTable/VoteButton.vue:
```diff
@@ -48,6 +48,8 @@ const vote = computed(() =>
 	}),
 )
 
+const chosenRank = JSON.parse(pollStore.configuration.chosenRank)
```
This is where we should go, yet i reduce the JSON parsing in fewer frontend files, given i don't handle code that does backend tweaks.

Commit:
chosenRank string nullable

Signed-off-by: philippe lhardy <philippe.lhardy@astrolabe.coop>

artlog commented on lib/Db/Poll.php:
```diff
@@ -148,6 +151,7 @@ class Poll extends EntityWithUser implements JsonSerializable {
 	protected string $access = '';
 	protected int $anonymous = 0;
 	protected int $allowMaybe = 0;
+	protected string $chosenRank = '';
```
resolved as requested

artlog commented on lib/Db/Poll.php:
```diff
@@ -183,6 +187,7 @@ public function __construct() {
 		$this->addType('anonymous', 'integer');
 		$this->addType('allowComment', 'integer');
 		$this->addType('allowMaybe', 'integer');
+		$this->addType('chosenRank', 'string');
```
done.

artlog committed:
Merge branch 'main' into variant_generic_unified_main

Signed-off-by: philippe lhardy <pl@artisanlogiciel.net>

artlog commented:
@dartcafe : after refereshing my environment with last merge i get this error :

```
Error: Typed property OCA\Polls\Db\V2\DbManager::$schema must not be accessed before initialization in /var/www/html/custom_apps/polls/lib/Db/V2/TableManager.php:466
Stack trace:
#0 /var/www/html/custom_apps/polls/lib/Migration/Version080301Date20250822153002.php(84): OCA\Polls\Db\V2\TableManager->fixNullishShares()
#1 /var/www/html/lib/private/DB/MigrationService.php(499): OCA\Polls\Migration\Version080301Date20250822153002->preSchemaChange(Object(OC\Migration\SimpleOutput), Object(Closure), Array)
#2 /var/www/html/lib/private/DB/MigrationService.php(395): OC\DB\MigrationService->executeStep('080301Date20250...', false)
#3 /var/www/html/lib/private/legacy/OC_App.php(694): OC\DB\MigrationService->migrate()
#4 /var/www/html/lib/private/Updater.php(325): OC_App::updateApp('polls')
#5 /var/www/html/lib/private/Updater.php(236): OC\Updater->doAppUpgrade()
#6 /var/www/html/lib/private/Updater.php(100): OC\Updater->doUpgrade('31.0.6.2', '31.0.6.2')
#7 /var/www/html/core/Command/Upgrade.php(192): OC\Updater->upgrade()
#8 /var/www/html/3rdparty/symfony/console/Command/Command.php(326): OC\Core\Command\Upgrade->execute(Object(Symfony\Component\Console\Input\ArgvInput), Object(Symfony\Component\Console\Output\ConsoleOutput))
#9 /var/www/html/3rdparty/symfony/console/Application.php(1078): Symfony\Component\Console\Command\Command->run(Object(Symfony\Component\Console\Input\ArgvInput), Object(Symfony\Component\Console\Output\ConsoleOutput))
#10 /var/www/html/3rdparty/symfony/console/Application.php(324): Symfony\Component\Console\Application->doRunCommand(Object(OC\Core\Command\Upgrade), Object(Symfony\Component\Console\Input\ArgvInput), Object(Symfony\Component\Console\Output\ConsoleOutput))
#11 /var/www/html/3rdparty/symfony/console/Application.php(175): Symfony\Component\Console\Application->doRun(Object(Symfony\Component\Console\Input\ArgvInput), Object(Symfony\Component\Console\Output\ConsoleOutput))
#12 /var/www/html/lib/private/Console/Application.php(187): Symfony\Component\Console\Application->run(Object(Symfony\Component\Console\Input\ArgvInput), Object(Symfony\Component\Console\Output\ConsoleOutput))
#13 /var/www/html/console.php(87): OC\Console\Application->run(Object(Symfony\Component\Console\Input\ArgvInput))
#14 /var/www/html/occ(33): require_once('/var/www/html/c...')

```

dartcafe commented:
Yep. I accessed thew table schema to check for the existence of a column, but forgot to initialize it.

artlog committed:
Merge branch 'main' into variant_generic_unified_main

artlog commented:
@dartcafe what is status of this, will it be integrated or what is expected from me ?
BTW i am at berlin community nextcloud conference until 2th october 2025, if you are around.. i will get some community help here too.

dartcafe commented:
I had no change to look into it, since the problem around the migration got priority and my time is limited atm due to some personal tasks and work.
